File: Csharp/Behavioral/State/Show/Show/Context.cs
namespace Show
{
    public class Context
    {
        private IState _state;

        public Context()
        {
            _state = new SleepState();
        }

        public void SetState(IState state)
        {
            _state = state;
        }

        public void Request()
        {
            _state.Handle(this);
        }
    }
}

File: JavaScript/Behavioral/State/Example1/Context.js
// Context.js

// Context maintains a reference to the current state
class Context {
    constructor(state) {
        this.state = state; // start with an initial state
    }

    setState(state) {
        this.state = state; // update the current state
        console.log(`Context: changed state to ${state.constructor.name}`);
    }

    request() {
        this.state.handle(this); // delegate to the current state's handle
    }
}

module.exports = Context;
